0000-qiskit-pulse-schedule-relative-timing.md

```diff
@@ -0,0 +1,103 @@
+
+Furthermore, explicit timing of instructions will become confusing with the introduction of control-flow to quantum programs as it would become difficult to program experiments containing loops that are not unrollable.
+
+We should move to a model where instructions are explicitly scheduled with variable-duration `Delay`s (think NOPs) of a given duration. This will:
+
+ - Make appending two schedules a much simpler operation as the dependency of instructions will be specified by their order on operands
+ - This will allow a simple transformation to a DAG representation of a basic pulse schedule block
+ - Have positive performance impacts.
+ - Have implications for the current pulse API.
+
+## User Benefit
+ - Pulse users (primarily, experimental physicists interested in using IBM backends) should get a performance improvement.
+ - When we begin implementing control flow around Schedules, we will find this new implementation much easier to work with. This benefits frontend and backend developers, which in turn will give our end users new features sooner.
+
+## Design Proposal
+
+### Implementation changes
+
+Here we will use the following notation, where `self` is the `Schedule` initiating the operation in question:
+
+ - |C| : the total number of channels in `self`
+ - N<sub>C</sub> : the number of instructions on channel C within `self`
+
+| Method        | Runtime        |  Change          |
+|---------------|----------------|------------------|
+| `append(self, schedule)` | O(\|C\|) | Previously, `append` would add `schedule` to `self` with relative times preserved, at time set by `self.duration({schedule.channels}.intersection({self.channels}))`. The new method will simply extend the instructions in `self` with those in `schedule` on a per channel basis. |
+|`insert(self, time, schedule)`| O(N<sub>new</sub> * N<sub>C</sub>) | This will be slow generally, but it can be implemented with the same runtime as `append` when the `time` is greater than `self.duration(schedule.channels)`. |
+
+
+## Detailed Design
+
+The `Schedule` will have an internal attribute, `_insts_by_chan`:
+
+```
+_insts_by_chan: Dict[Channel, List[ScheduleNode]]
+```
+
+**Example**
+```
+# sched._insts_by_chan
+{
+    DriveChannel(0):
+        ('instructions'=[Delay(duration=10, ...),
+                         Play(duration=100, ...),
+                         Play(duration=20, ...)],
+         'duration'=130),
+     DriveChannel(1):
+        ('instructions'=[Delay(duration=110, ...),
+                         Play(duration=20, ...)],
+         'duration'=130),
+    ...
+}
```

Comment:
This seems to be the main takeaway of this RFC, so we should settle on it.

So the 130 number is here for performance right? Even though it could be computed by summing all durations on that channel. I'm good with that.

However, we should think about what a conditional will look like. If I don't know an instruction's delay, due to a branch or conditional, how does this datastructure work?

Reply:
A `Schedule` is a "basic block", so conditionals should live on top of this. An alt approach (which I should list in this RFC) would be to make this explicit and allow absolute timing within a basic block even when we have control flow. I find that misleading (too "special case"-y), but, at the heart of it, they're equivalent.

Reply:
I think Thomas' new RFC touches on this